### ISSUE TITLE
Typechecking of context doesn't work with custom directives

### ISSUE THREAD

**Issue by majellin24** [angular/vscode-ng-language-service#1158]
I'm not sure if this is appropriate but this is basically just a duplicate of https://github.com/angular/vscode-ng-language-service/issues/650

Discussion within that issue seemed to suggest that the issue would be resolved using the ivy language service, however, I cant seem to get that to work and I'm just wondering about the status of the work toward resolving that issue or if there are additional steps I need to take to get it working.  I've tried using the experimental ivy setting and leaving that option unchecked with no success.

In my scenario, the context is just
```
export class TemplateContext<T, U extends T[] = T[]> {
  constructor(public $implicit: T, public ref: U){}
}

@Directive({
  selector: '[spTemplateContent]'
})
export class TemplateContentDirective<T, U extends T[] = T[]> {

  @Input()
  spTemplateContentFrom: U

  constructor(public content: TemplateRef<TemplateContext<T>>) {
  }
  static ngTemplateContextGuard<T, U extends T[] = T[]>(dir: TemplateContentDirective<T, U>, ctx: any):
      ctx is TemplateContext<T, U> {
      return true;
    }
}
```
and using 
```
<... *myDirective="let ctx; from: contentList" ...>
    {{ctx}}
</...>
```
shows the type of "ctx" to be "any" but I would like it to show the type of an element of "contentList."  For instance, if contentList is a "MyCoolClass[]" then I would like ctx to show as type "MyCoolClass"

**Comment by atscott** [comment]
The types shown by Angular Language Service with Ivy reflect what you've configured the compiler to interpret. Have you turned on `strictTemplates` or `fullTemplateTypeCheck` at the very least in the [angular compiler options](https://angular.io/guide/angular-compiler-options)? If not, Angular does not type check the bodies of templates.

**Comment by majellin24** [comment]
Yes, `fullTemplateTypeCheck` is true

**Comment by atscott** [comment]
@majellin24 - Can you provide a reproduction in a GitHub repository so I can investigate?

**Comment by majellin24** [comment]
@atscott sure, I'll work on that.  One more thing which might be relevant is that when I have "Experimental-ivy" checked in vscode's extension settings, I am also experiencing what is mentioned in https://github.com/angular/vscode-ng-language-service/issues/1155

I'm wondering if this might be a configuration issue for me.  This might be a dumb question but I looked for the answer and could not find it.  Does npm package for `@angular/language-service` need to match the vscode extension version?

**Comment by atscott** [comment]
@majellin24 You don't need to install the `@angular/language-service` package. This is bundled with the extension.

**Comment by kyliau** [comment]
> I'm wondering if this might be a configuration issue for me. This might be a dumb question but I looked for the answer and could not find it. Does npm package for `@angular/language-service` need to match the vscode extension version?

@majellin24 If you could provide the logs it'll give us better insight into your config / project setup. The `angularCompilerOptions` are logged there.

**Comment by atscott** [comment]
Closing as duplicate of #1155.

@majellin24 please add `"strictInputTypes": true` to your compiler options at a minimum to get this working. We recommend `"strictTemplates": true` for the most complete compiler information but understand that projects have to go through some internal cleanup to make this happen.

**Comment by majellin24** [comment]
@atscott I'm comfortable with this being closed as a duplicate, although I will try to get a minimal reproduction and add the language service logs from that.

I just want to clarify that you would expect https://github.com/angular/vscode-ng-language-service/issues/650 to work properly now if I get my https://github.com/angular/vscode-ng-language-service/issues/1155 type issue resolved?

Thanks

**Comment by atscott** [comment]
@majellin24 Yes, with the Ivy Language Service enabled, #650 will work. The fix for #1155 is to have either `"strictTemplates": true` in the `angularCompilerOptions` or both `"fullTemplateTypeCheck": true` and `"strictInputTypes": true`.

**Comment by atscott** [comment]
@majellin24 Also, don't worry about the reproduction and logs. I think we've identified the root cause. Of course, do feel free to provide them if the provided solution doesn't work for you.

**Comment by majellin24** [comment]
@atscott sorry to keep commenting but actually just setting `"strictInputTypes": true` solved both issues for me (#650 and #1155) so I wanted to let you know.  Thanks a bunch!
